wdl-format/src/lib.rs

```diff
@@ -0,0 +1,215 @@
+use wdl_ast::AstToken;
+use wdl_ast::Comment;
+use wdl_ast::Diagnostic;
+use wdl_ast::Direction;
+use wdl_ast::Document;
+use wdl_ast::Ident;
+use wdl_ast::SyntaxElement;
+use wdl_ast::SyntaxKind;
+use wdl_ast::SyntaxNode;
+use wdl_ast::Validator;
+use wdl_ast::Version;
+use wdl_ast::VersionStatement;
+
+mod comments;
+mod formatter;
+mod import;
+mod metadata;
+mod task;
+mod v1;
+mod workflow;
+
+use comments::format_inline_comment;
+use comments::format_preceding_comments;
+use formatter::Formatter;
+
+/// Newline constant used for formatting on windows platforms.
+#[cfg(windows)]
+pub const NEWLINE: &str = "\r\n";
+/// Newline constant used for formatting on non-windows platforms.
+#[cfg(not(windows))]
+pub const NEWLINE: &str = "\n";
+/// String terminator constant used for formatting.
+const STRING_TERMINATOR: char = '"';
+
+/// A trait for elements that can be formatted.
+pub trait Formattable {
+    /// Format the element and write it to the writer.
+    fn format<T: std::fmt::Write>(
+        &self,
+        writer: &mut T,
+        formatter: &mut Formatter,
+    ) -> std::fmt::Result;
+}
+
+impl Formattable for Version {
+    fn format<T: std::fmt::Write>(
+        &self,
+        writer: &mut T,
+        _formatter: &mut Formatter,
+    ) -> std::fmt::Result {
+        write!(writer, "{}", self.as_str())
+    }
+}
+
+impl Formattable for VersionStatement {
+    fn format<T: std::fmt::Write>(
+        &self,
+        writer: &mut T,
+        formatter: &mut Formatter,
+    ) -> std::fmt::Result {
+        let mut preamble_comments = Vec::new();
+        let mut lint_directives = Vec::new();
+        let comment_buffer = &mut String::new();
+        for sibling in self.syntax().siblings_with_tokens(Direction::Prev).skip(1) {
+            match sibling.kind() {
+                SyntaxKind::Comment => {
+                    let comment = Comment::cast(
+                        sibling
+                            .as_token()
+                            .expect("Comment should be a token")
+                            .clone(),
+                    )
+                    .expect("Comment should cast to a comment");
+                    comment.format(comment_buffer, formatter)?;
+                    if comment_buffer.starts_with("#@") {
+                        lint_directives.push(comment_buffer.clone());
+                    } else {
+                        preamble_comments.push(comment_buffer.clone());
+                    }
+                    comment_buffer.clear();
+                }
+                SyntaxKind::Whitespace => {
+                    // Ignore
+                }
+                _ => {
+                    unreachable!("Unexpected syntax kind: {:?}", sibling.kind());
+                }
+            }
+        }
+
+        for comment in preamble_comments.iter().rev() {
+            write!(writer, "{}", comment)?;
+        }
+
+        // If there are preamble comments, ensure a blank line is inserted
```

Comment:
This was decided by me and @adthrasher a long time ago, and has been codified by the rule `PreambleWhitespace`: https://github.com/stjude-rust-labs/wdl/blob/main/wdl-lint/src/rules/preamble_whitespace.rs#L70

That rule predates `sprocket` and lint directives (`#@` comments). A stylistic choice I made in this PR is that lint directives should have a blank separating them from preamble comments (`##`) and occur _without_ a blank separating them from the version statement. Like so,
```
## # Title
## preamble

#@ except: Foo
#@ except: Bar
version 1.1
```
The above will result in lint diagnostics from `PreambleWhitespace`, but I thought that could be tweaked in a follow-up PR (it would involve just a few small changes) that would give us all an opportunity to voice our thoughts on that style choice.

IMO the part of the preamble that gets converted to markdown should have some separation from "everything else". Hence the above choices.

Reply:
Okay, sounds fine.